File: backend/app/models/simulation.py
from __future__ import annotations

from pydantic import BaseModel, Field


class FinancialImpact(BaseModel):
    posting_required: bool
    amount: float | None = None
    currency: str | None = None
    gl_accounts_affected: list[str] = Field(default_factory=list)
    cost_centers_affected: list[str] = Field(default_factory=list)


class WorkflowImpact(BaseModel):
    steps_triggered: list[str] = Field(default_factory=list)
    approvals_required: list[str] = Field(default_factory=list)
    notifications_sent: list[str] = Field(default_factory=list)


class SimulationResult(BaseModel):
    documents_affected: int
    financial: FinancialImpact
    workflow: WorkflowImpact
    risk_score: float = Field(..., ge=0.0, le=1.0, description="Composite risk 0-1")
    warnings: list[str] = Field(default_factory=list)
    blockers: list[str] = Field(
        default_factory=list, description="Issues that prevent execution"
    )
    reversible: bool = Field(default=True)
    simulation_notes: str = ""

File: frontend/src/types/simulation.ts
export interface FinancialImpact {
  posting_required: boolean
  amount: number | null
  currency: string | null
  gl_accounts_affected: string[]
  cost_centers_affected: string[]
}

export interface WorkflowImpact {
  steps_triggered: string[]
  approvals_required: string[]
  notifications_sent: string[]
}

export interface SimulationResult {
  documents_affected: number
  financial: FinancialImpact
  workflow: WorkflowImpact
  risk_score: number
  warnings: string[]
  blockers: string[]
  reversible: boolean
  simulation_notes: string
}

export interface SimulationResponse {
  action_tcode: string
  simulation: SimulationResult
  simulated_at: string
}

export interface SimulateRequest {
  tcode: string
  module: string
  document_id: string
  status?: string
  action_tcode: string
  parameters?: Record<string, unknown>
}
